### ISSUE TITLE
Document rationale for why short-circuiting operators don't fold RHS param values

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#20873]
I was surprised to find that the following conditional is not folded as I would have expected:

```chapel
config param flag = false;

if foo() && flag then   // *** conditional of interest ***
  bar();

writeln("OK!");

proc foo() {
  return true;
}

proc bar() param {
  compilerError("Shouldn't get here");
}
```

seemingly because of `flag`/`false` appearing as the second argument to `&&`.  If the two arguments are swapped, things work as you'd expect.

[edit: In the comments below, @daviditen points out, compellingly, that this is by design so that the first expression is evaluated for side effects; I've re-purposed this issue to be about noting the rationale for this in the language spec given how many of us assumed it would be folded]

[edit: See also #20496, which is related]

**Associated Test:**
[`test/param/shortCircuitParamRight.chpl`](
           https://github.com/chapel-lang/chapel/blob/main/test/param/shortCircuitParamRight.chpl
          ) #21332

**Comment by mppf** [comment]
@DanilaFe - IIRC you worked a bit on dyno's param folding. Could you check to see if this bug exists in our prototype dyno resolver?

**Comment by DanilaFe** [comment]
It sure does. I'll submit a fix PR.

**Comment by DanilaFe** [comment]
We do, however, have tests that _make sure_ folding on the right doesn't occur...
```C++
  // 2nd argument param should not be folded
  QualifiedType qt = resolveTypeOfXInit(context,
                                      "var a: bool; var x = a || true;");
  assert(!qt.isParam() && !qt.hasParamPtr());
  assert(qt.type() == BoolType::get(context, 0));
```

**Comment by bradcray** [comment]
Oh, interesting.  Best guess:  In general, the RHS of a short-circuiting operator shouldn't be evaluated unless the LHS suggests that it needs to be.  So maybe it was intentional to work hard not to evaluate the param if the RHS is an execution-time expression to ensure that the execution-time expression runs?  But given that all params need to be evaluated at compile-time before an execution-time expression runs, it still seems a little odd to suggest deferring the folding of the conditional, particularly since param conditionals can be so powerful.

Who wrote the test?

If we decided that this was correct behavior for reasons that I'm not picking up on, it seems as though we might want to issue a warning saying "This conditional won't be folded at compile-time, but potentially could be if you swapped the order of the arguments"?

**Comment by DanilaFe** [comment]
I believe @mppf wrote the test. I agree with your assessment, Brad.

**Comment by daviditen** [comment]
> Best guess: In general, the RHS of a short-circuiting operator shouldn't be evaluated unless the LHS suggests that it needs to be.

Yes, it was designed to do exactly this. The LHS must be evaluated before the RHS is considered. The LHS might have side effects that would be lost if the RHS caused the whole expression to be folded. The RHS is still a param, but the expression as a whole is not.

**Comment by DanilaFe** [comment]
I can make code like this issue a warning on the Dyno side, then.

**Comment by bradcray** [comment]
It'd be good to get some documentation into the spec indicating this behavior as well (if we don't already have it) and rationalizing it.

**Comment by daviditen** [comment]
I think this section describes the expected behavior fairly well, in particular the `if isTrue(a) then isTrue(b) else false` example.

https://chapel-lang.org/docs/master/language/spec/expressions.html#the-logical-and-operator

**Comment by bradcray** [comment]
If you're arguing that we should add overloads of `*` that takes param 0's and fold them, I completely agree with you and would fully support your opening that issue and/or PR.

> I also don't think what's there already is unclear.

I'm not saying the current text is unclear, I'm just saying that sometimes even when two clear rules (short-circuiting operator definitions, conditional folding) combine and result in a reasonable behavior, that doesn't mean that the intersection of the behavior is self-evident.  In this case I'd say that it wasn't given that I opened this issue, nobody proactively pointed out my mistake, Michael seemed to agree with my thinking, Daniel and his implementation seemed to as well, and it took this conversation for us to realize that I was mistaken.  That seems well worth having a note pointing out that even though a human may look at `myVar && false` or `myVar || true` and think "These can obviously be folded into the params `false`/`true`, they'd be wrong due to the need to evaluate the LHS for side-effects."

(it also raises an interesting question about whether the compiler could fold the cases where it proves the LHS is side-effect free, where I'd expect "yes", but I'm not as worried about that personally...)

**Comment by mppf** [comment]
I think tricky cases like this make great examples to explain how the (hopefully) clear rules are interacting.

**Comment by bradcray** [comment]
I returned to this issue for the first time in awhile, updating its title and description to reflect my sense that we should update the documentation to capture the rationale here.

That said, reading #20496 and thinking about this issue a bit more, I'm back to being unconvinced that the current behavior is correct.  Specifically, I still buy @daviditen's assertion that we have to evaluate the LHS for side-effects before evaluating the RHS.  But at the same time, there is no way that the result of the conditional expression can be `true`, whatever the LHS expression is.  Therefore, it still seems suspect that we'd bother resolving the `then` conditional of the loop body, knowing that it can never execute.  Put another way, though we have to evaluate the conditional expression at execution time, we still know what the param result must be.  #20496 argues that we should compute the `param` but still evaluate the LHS for side-effects, and that sounds pretty reasonable (and less surprising than today's behavior) to me.  Any counterarguments?

**Comment by bradcray** [comment]
In the comment just above, I argued that even though we should evaluate the first argument for side-effects, maybe we should fold the body of the conditional anyway and asked for counterarguments to that proposal.

A counterargument that was brought up in a deep-dive today is that it can be useful to have the compiler stably resolve a given branch—whether or not it's executed—in order to catch any errors that might exist in that branch (e.g., typos/thinkos that parse but result in resolution errors or the like).

But I think a counterargument to that counterargument might be that if I've intentionally written the conditional so that it doesn't make sense when the param conditional would fold, then I'm losing power in the language if the compiler won't fold it for me.

Saying that with an example, today the language guarantees that the following conditional will be squashed when `param` is true:

```chapel
config type t = real(32);
config param squash = (numBits(t) != 32);
var x: t = …;

if squash {
  var u: uint(32) = x.transmute(uint(32));  // this code only makes sense for 32-bit types
  …
}
```

If we were to say that it would never squash:

```chapel
if mySideEffectingExpr && squash {
  var u: uint(32) = x.transmute(uint(32));
}
```

unless I rewrote it as:

```chapel
param squash2 = mySideEffectingExpr && squash;
if squash2 {
  var u: uint(32) = x.transmute(uint(32));
}
```

then it feels a bit like we're forcing the user to jump through some hoops to get the behavior they wanted.

That is the _power_ of having a language that can fold conditionals as a tool for programmers to squash code seems at odds with the desire for _consistency_ in resolving as much as possible (while also introducing an arguable inconsistency in how simply the conditionals need to be written).

Back on the other foot, I think this whole question boils down to whether we consider `sideEffectingNonParamExpr && paramBool` to be a `param` expression or not.  If we do, then I think we have to fold such conditionals since we should fold `if paramBoolExpr` conditionals consistently whatever the source of the param boolean expression is.  But if we don't, then we have a good excuse to treat this conditional differently.  And I think the potential for side effects is a good argument for not considering it a `param`.

If the language were (someday) rich enough to be able to distinguish side-effecting/pure expressions from non-, then I could imagine an argument that `pureNonParamExpr && paramBool` should be a param, but we're not in that world today.

**Comment by bradcray** [comment]
> And I think the potential for side effects is a good argument for not considering it a param.

Though, contradicting myself again, we don't evaluate `param`-returning functions for side effects:

```chapel
use Types;

proc bits32(x) param {
  if numBits(x.type) == 32 {
    writeln("Got 32");
    return true;
  } else {
    writeln("Got something else");
    return false;
  }
}

var i: int;
var j: int(32);
param res = bits32(i);
param res2 = bits32(j);
writeln((res, res2));
```

generates:

```
(false, true)
```

**Comment by bradcray** [comment]
Still spinning apparently:

I find myself wanting to say that issues like this boil down to the question “Does `expr1 op expr2` generate a param if `expr1` and/or `expr2` is a `param`?” where I'd like to say that the answer should be independent of the specific values of `expr1` and `expr2`.  So, since `expr < 42` can’t generally be a param for integral `expr` then `expr < 0` shouldn’t be either.  Then we can say that any `if paramExpr` is folded or not depending on whether the value is `param false` or `true`.

But then I realize that `true && myExpr` vs. `false && myExpr` are cases where the `param`-ness of the result depends on the value of first argument, contradicting the above.  Yet they're also cases that we've been happy to have fold.  So do we consider them to be a special case?  (and for `myExpr && true` or `myExpr && false` to not be special cases?)  Or is the instinct of the previous paragraph just completely wrong?
